packages/flutter/lib/src/material/tab_controller.dart

```diff
@@ -0,0 +1,201 @@
+  }
+
+  /// An animation whose value represents the current position of the [TabBar]'s
+  /// selected tab indicator as well as the scrollOffsets of the [TabBar]
+  /// and [TabBarView].
+  ///
+  /// The animation's value ranges from 0.0 to [length] - 1.0. After the
+  /// selected tab is changed, the animation's value equals [index]. The
+  /// animation's value can be [offset] by +/- 1.0 to reflect [TabBarView]
+  /// drag scrolling.
+  final AnimationController _animationController;
+  Animation<double> get animation => _animationController.view;
+
+  /// The total number of tabs. Must be greater than one.
+  final int length;
+
+  void _changeIndex(int value, { Duration duration, Curve curve }) {
+    assert(value != null);
+    assert(value >= 0 && value < length);
+    assert(duration == null ? curve == null : true);
+    if (value == _index)
+      return;
+    _indexIsChanging = true;
+    _previousIndex = index;
+    _index = value;
+    if (duration != null) {
+      _animationController
+        ..animateTo(_index.toDouble(), duration: duration, curve: curve).then((_) {
+          _indexIsChanging = false;
+          notifyListeners();
+        });
+    } else {
+      _indexIsChanging = true;
+      _animationController.value = _index.toDouble();
+      _indexIsChanging = false;
+      notifyListeners();
+    }
+  }
+
+  /// The index of the currently selected tab. Changing the index also updates
+  /// [previousIndex], sets the [animation]'s value, resets [indexIsChanging]
+  /// to false, and notifies listeners.
+  ///
+  /// To change the currently selected tab and play the [animation] use [animateTo].
+  int get index => _index;
+  int _index;
+  set index(int value) {
+    _changeIndex(value);
+  }
+
+  /// The index of the previously selected tab. Initially the same as [index].
+  int get previousIndex => _previousIndex;
+  int _previousIndex;
+
+  /// True while we're animating from [previousIndex] to [index].
+  bool get indexIsChanging => _indexIsChanging;
+  bool _indexIsChanging = false;
+
+  /// Immediately sets [index] and [previousIndex] and then plays the
+  /// [animation] from its current value to [index].
+  ///
+  /// While the animation is running [indexIsChanging] is true. When the
+  /// animation completes [offset] will be 0.0.
+  void animateTo(int value, { Duration duration: kTabScrollDuration, Curve curve: Curves.ease }) {
+    _changeIndex(value, duration: duration, curve: curve);
+  }
+
+  /// The difference between the [animation]'s value and [index]. The offset
+  /// value must be between -1.0 and 1.0.
+  ///
+  /// This property is typically set by the [TabBarView] when the user
+  /// drags left or right. A value between -1.0 and 0.0 implies that the
+  /// TabBarView has been dragged to the left. Similarly a value between
+  /// 0.0 and 1.0 implies that the TabBarView has been dragged to the right.
+  double get offset => _animationController.value - _index.toDouble();
+  set offset(double value) {
```

Comment:
I'd use a different variable name here because `value` is confusing with `_animationController.value`.  Maybe `newOffset`?